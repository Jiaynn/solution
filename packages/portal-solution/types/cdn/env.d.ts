/**
 * @file env config
 * @author linchen <linchen@qiniu.com>
 */

declare interface OEMConfig {
  logo: string
  vendor: number
  favicon: string
  title: string
  copyright: string
  lang: 'en' | 'cn'
  domainHosting: boolean                  // 域名托管功能
  financial: boolean                      // 财务中心
  hideLoginTitle: boolean                 // 隐藏登录页面的 title
  hideCreateDomain: boolean               // 隐藏创建域名的操作入口
  hideUpdateDomain: boolean               // 隐藏更新域名的操作入口
  hideSubAccountDomainListCname: boolean  // 对子用户隐藏域名列表页的 cname
  hideSubAccountDomainDetailCname: boolean // 对子用户隐藏域名配置详情页的 cname
  hideSubAccountDomainOperations: boolean // 对子用户隐藏域名的操作 (只保留统计操作)
  hideSubAccountCreateDomain: boolean     // 对子用户隐藏创建域名的操作
  hideSubAccountGeoForeign: boolean       // 对子用户禁用海外覆盖范围
  hideSubAccountGeoGlobal: boolean        // 对子用户禁用全球覆盖范围
  hideSubAccountStatsUsagePeakAverage: boolean    // 对子用户隐藏平均峰值带宽
  hideSubAccountStatsUsagePeak95Average: boolean  // 对子用户隐藏平均 95 峰值带宽
  statisticsDateRangeDays: number // 数据统计默认时间范围, 最近的天数
  statisticsFreq: string // 数据统计粒度 '1day' | '1hour' | '5min'
}

declare let BUILD_TARGET: 'qiniu' | 'oem'
declare let OEM_CONFIG: OEMConfig
declare let APP_VERSION: string
