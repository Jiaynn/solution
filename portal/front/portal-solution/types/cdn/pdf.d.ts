declare let pdfExport: string

declare module '*.pdf' {
  export = pdfExport
}
