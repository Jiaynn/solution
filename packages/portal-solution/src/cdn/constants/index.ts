// 启用动态加速工单地址
export const enableDynamicURL = 'https://support.qiniu.com/tickets/new/form?category=%E5%85%B6%E4%BB%96%E7%B1%BB%E5%92%A8%E8%AF%A2&space=%E8%9E%8D%E5%90%88%20CDN'

// 证书配置提交工单 URL
export const certConfigHelpURL = 'https://support.qiniu.com/tickets/new/form?category=%E8%AF%81%E4%B9%A6%E9%97%AE%E9%A2%98&space=CDN'
