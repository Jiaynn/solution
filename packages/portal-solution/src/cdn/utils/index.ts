import { values } from 'lodash'
import { ReactNode } from 'react'
import { makeCancelled, UnknownException } from 'qn-fe-core/exception'

export function isPromise(obj: any): obj is Promise<any> {
  if (!obj) {
    return false
  }
  if (typeof obj === 'object' || typeof obj === 'function') {
    if (typeof obj.then === 'function') {
      return true
    }
  }
  return false
}

export interface IAntdStyleMethodOptions {
  title?: ReactNode
  content?: ReactNode
  onOk?: (...args: any[]) => void
  onCancel?: (...args: any[]) => void
}

export type AntdStyleMethod<T extends IAntdStyleMethodOptions> = (options: T) => void

export type PromiseStyleMethod<T, P> = (options: T) => Promise<P>

// 把 antd 中如 Modal.confirm 风格（onCancel, onOk）的接口转换成 promise 接口
export function antdToPromise<
  T extends IAntdStyleMethodOptions,
  M extends AntdStyleMethod<T>,
  P
>(method: M): PromiseStyleMethod<T, P> {
  return (options: T) => new Promise((resolve, reject) => {
    const onCancel = (...args: any[]) => {
      if (options.onCancel) {
        options.onCancel(...args)
      }
      reject(makeCancelled())
    }
    const onOk = (...args: any[]) => {
      if (options.onOk) {
        options.onOk(...args)
      }
      resolve(args[0])
    }
    method({
      ...(options as any),
      onCancel,
      onOk
    })
  })
}

function getLexicValue(char: string) {
  return char ? char.charCodeAt(0) : -1
}

export function lexicCompare(strA: string, strB: string) {
  for (let i = 0, len = Math.max(strA.length, strB.length); i < len; i++) {
    const vA = getLexicValue(strA.charAt(i))
    const vB = getLexicValue(strB.charAt(i))
    if (vA !== vB) {
      return vA - vB
    }
  }
  return 0
}

export function valuesOfEnum<T extends object>(target: T): Array<T[keyof T]> {
  const fakeKeys = values(target).filter(value => typeof value !== 'string').map(String)
  const realKeys = Object.keys(target).filter(key => fakeKeys.indexOf(key) === -1)

  return (realKeys as Array<keyof T>).map(key => target[key])
}

export function assertUnreachable(): never
export function assertUnreachable(v: never, message?: string): never
export function assertUnreachable(v?: never, message = "Didn't expect to get here") {
  throw new UnknownException(message + v)
}

/**
 * 过滤数组的 “空” 值，其中空值指的是：
 * - null
 * - undefined
 */
export function nonEmptyArray<T>(data: Array<T | undefined | null>): T[] {
  return data.filter((it): it is T => it != null)
}

export function booleanPredicate<T>(value: T): value is Exclude<T, false | null | undefined | '' | 0> {
  return Boolean(value)
}

export function assert(condition: any, msg?: string): asserts condition {
  if (!condition) {
    throw Error(msg)
  }
}
