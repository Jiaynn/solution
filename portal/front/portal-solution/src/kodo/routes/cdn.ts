/**
 * @file CDN route TODO: 站内外链与内链分开
 * @author yinxulai <me@yinxulai.com>
 */

// CDN 概览
export function getCDNOverviewPath() {
  return '/cdn/overview'
}

// CDN 业务的域名创建
export function getCDNCreateDomainPath(bucketName: string) {
  const query = bucketName ? `?bucket=${encodeURIComponent(bucketName)}&fixBucket` : ''
  return `/cdn/domain/create${query}`
}

// CDN 业务的域名列表
export function getCDNDomainListPath() {
  return '/cdn/domain'
}

export function getCDNDomainDetailPath(domain: string) {
  return `/cdn/domain/${domain}`
}

export function getCDNCreateBucketDomainPath(bucketName: string) {
  return `/cdn/domain/create?bucket=${bucketName}&fixBucket`
}

export function getCDNRefreshPath() {
  return '/cdn/refresh-prefetch'
}

export function getCDNRefreshLogsPath() {
  return '/cdn/refresh-prefetch?tab=showLogs'
}
