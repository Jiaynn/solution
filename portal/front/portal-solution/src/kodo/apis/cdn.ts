/**
 * @file config api
 * @description region 相关的接口
 * @author lizhifeng <lizhifeng@qiniu.com>
 */

import autobind from 'autobind-decorator'
import { injectable } from 'qn-fe-core/di'
import { ApiException } from 'qn-fe-core/client'

import { ProxyClientV2 } from 'portal-base/common/apis/proxy'

import { proxy } from 'kodo/constants/apis'

interface CdnBaseResponse {
  code: number
  error: string
  [k: string]: any
}

export interface RefreshCdnSurplus extends CdnBaseResponse {
  urlQuotaDay: number // url 刷新当日限额
  urlSurplusDay: number // url 刷新当日余量
  dirQuotaDay: number // 目录刷新当日限额
  dirSurplusDay: number // 目录刷新当日余量
  quotaDay: number // 预取当日限额
  surplusDay: number // 预取当日剩余量
}

interface RefreshCdnOptions {
  product: 'cdn'
  urls: string[]
}

export class CdnApiException extends ApiException {
  constructor(
    public code: number,
    public payload: CdnBaseResponse,
    message?: string,
    cause?: unknown
  ) {
    super(
      'RefreshPrefetchApiException',
      code,
      message,
      undefined,
      cause
    )
  }
}

const messageMap = new Map<number, string>([
  [400031, '无效的 URL'],
  [400034, '刷新域名个数超过限制'],
  [400037, 'URL 已在队列中，无需重复提交'],
  [400039, '请求次数过多']
])

function produceResult<T extends CdnBaseResponse>(payload: T) {
  if (payload && typeof payload === 'object' && 'code' in payload && payload.code !== 200) {
    const message = messageMap.get(payload.code) || payload.error
    throw new CdnApiException(payload.code, payload, message)
  }

  return payload
}

@autobind
@injectable()
export class CdnApis {
  constructor(
    private proxyClient: ProxyClientV2
  ) {}

  getRefreshCdnSurplus(): Promise<RefreshCdnSurplus> {
    return this.proxyClient.get<RefreshCdnSurplus>(proxy.refreshCdnSurplus, {})
      .then(produceResult)
  }

  async refreshCdn(params: RefreshCdnOptions): Promise<void> {
    await this.proxyClient.post<CdnBaseResponse>(proxy.refreshCdn, params)
      .then(produceResult)
  }
}
