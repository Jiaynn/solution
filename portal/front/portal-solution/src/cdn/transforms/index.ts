import { QueryParams } from 'qn-fe-core/utils'

export function atoi(numberText: string, valueForInvalidInput = 0): number {
  const parsed = parseInt(numberText, 10)
  return Number.isNaN(parsed) ? valueForInvalidInput : parsed
}

export function atof(numberText: string, valueForInvalidInput = 0): number {
  const parsed = parseFloat(numberText)
  return Number.isNaN(parsed) ? valueForInvalidInput : parsed
}

export function getIntValue(e: React.FormEvent<any>, valueForInvalidInput = 0) {
  return atoi((e.target as any).value, valueForInvalidInput)
}

export function getFloatValue(e: React.FormEvent<any>, valueForInvalidInput = 0) {
  return atof((e.target as any).value, valueForInvalidInput)
}

export function trimAndFilter(lines: string[]): string[] {
  return (lines || []).map(
    line => line.trim()
  ).filter(
    line => !!line
  )
}

export function splitLines(linesStr: string | null | undefined): string[] {
  if (!linesStr) {
    return []
  }
  return linesStr.split('\n')
}

export function filterQuery(query?: QueryParams) {
  if (!query) {
    return
  }
  return Object.keys(query).reduce(
    (filtered, key) => (
      (query || {})[key]
      ? { ...filtered, [key]: (query || {})[key] }
      : filtered
    ), {} as QueryParams
  )
}

export function humanizeOnoff(isOn: boolean) {
  return isOn ? '开启' : '关闭'
}
