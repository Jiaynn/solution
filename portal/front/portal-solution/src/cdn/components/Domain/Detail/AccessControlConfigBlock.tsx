
import React from 'react'
import { useInjection } from 'qn-fe-core/di'
import { Iamed } from 'portal-base/user/iam'
import { ICertInfo } from 'portal-base/certificate'

import { shouldForbidModifyAccessControl } from 'cdn/transforms/domain'
import { humanizeRefererType } from 'cdn/transforms/domain/referer'
import { humanizeIpACLType } from 'cdn/transforms/domain/ip-acl'

import timeRefererDoc from 'cdn/docs/time-referer.pdf'

import IamInfo from 'cdn/constants/iam-info'
import { refererTypes, ipACLTypes, Platform } from 'cdn/constants/domain'

import HelpLink from 'cdn/components/common/HelpLink'

import { IDomainDetail } from 'cdn/apis/domain'

import BlockTitle from './BlockTitle'
import ConfigTable from './ConfigTable'
import ConfigureButton from './ConfigureButton'
import ConfigStatus from './ConfigStatus'

function getRefererText(domain: IDomainDetail) {
  const { refererType, nullReferer, refererValues } = domain.referer
  switch (refererType) {
    case refererTypes.empty:
      return <ConfigStatus enabled={false} />
    case refererTypes.black:
    case refererTypes.white: {
      const refererTypeText = humanizeRefererType(refererType)
      const nullRefererText = nullReferer ? '允许空 Referer' : '不允许空 Referer'
      return `${refererTypeText} ${refererValues.length} 个，${nullRefererText}`
    }
    default:
      return '未知'
  }
}

function getIpACLText(domain: IDomainDetail) {
  switch (domain.ipACL.ipACLType) {
    case ipACLTypes.empty:
      return <ConfigStatus enabled={false} />
    case ipACLTypes.black:
    case ipACLTypes.white: {
      const ipACLTypeText = humanizeIpACLType(domain.ipACL.ipACLType)
      return `${ipACLTypeText} ${domain.ipACL.ipACLValues.length} 个`
    }
    default:
      return '未知'
  }
}

export default function AccessControlConfigBlock(props: {
  domain: IDomainDetail
  loading: boolean
  certInfo?: ICertInfo
  handleConfigureReferer: () => void
  handleConfigureTimeReferer: () => void
  handleConfigureBsAuth: () => void
  handleConfigureIpACL: () => void
}) {
  const {
    domain,
    loading,
    certInfo,
    handleConfigureReferer,
    handleConfigureTimeReferer,
    handleConfigureBsAuth,
    handleConfigureIpACL
  } = props
  const iamInfo = useInjection(IamInfo)

  const isDynamic = domain.platform === Platform.Dynamic

  const refererInfo = {
    name: 'Referer 防盗链',
    desc: '配置 Request ConfigurationHeader 中 referer 黑白名单，从而限制访问来源。',
    value: getRefererText(domain),
    configureHandler: handleConfigureReferer,
    iamAction: () => iamInfo.iamActions.UpdateReferer
  }

  const timeRefererInfo = {
    name: '时间戳防盗链',
    desc: (
      <p>
        设置密钥，配合签名过期时间来控制资源内容的访问时限。
        <HelpLink oemHref={timeRefererDoc} href="https://support.qiniu.com/hc/kb/article/195128/" />
      </p>
    ),
    value: <ConfigStatus enabled={domain.timeACL.enable} />,
    configureHandler: handleConfigureTimeReferer,
    iamAction: () => iamInfo.mustCdnIamActions().UpdateTimeACL
  }

  const bsauthInfo = {
    name: '回源鉴权',
    desc: (
      <p>
        配置回源验证策略，对 CDN 接收到的访问请求进行鉴权，适用于对防盗链有很高实时性要求的场景。
        <HelpLink href="https://developer.qiniu.com/fusion/manual/3930/back-to-the-source-authentication" />
      </p>
    ),
    value: <ConfigStatus enabled={domain.bsauth && domain.bsauth.enable} />,
    configureHandler: handleConfigureBsAuth,
    iamAction: () => iamInfo.mustCdnIamActions().UpdateBSAuth
  }

  const ipInfo = {
    name: 'IP 黑白名单',
    desc: (
      <p>
        允许或者禁止某些 IP 或 IP 段的访问，帮助您解决恶意 IP 盗刷、攻击等问题。
      </p>
    ),
    value: getIpACLText(domain),
    configureHandler: handleConfigureIpACL,
    iamAction: () => iamInfo.iamActions.UpdateIpACL
  }

  const timeAndBsauthInfo = !isDynamic ? [timeRefererInfo, bsauthInfo] : []

  const infoList = [
    refererInfo,
    ...timeAndBsauthInfo,
    ipInfo
  ]

  const renderOperations = (_: unknown, record: any) => (
    <Iamed actions={[record.iamAction()]}>
      <ConfigureButton
        shouldForbid={shouldForbidModifyAccessControl(domain, certInfo)}
        onClick={record.configureHandler}
      >修改配置</ConfigureButton>
    </Iamed>
  )

  return (
    <section className="content-block">
      <BlockTitle>访问控制</BlockTitle>
      <ConfigTable
        configList={infoList}
        loading={loading}
        renderOperations={renderOperations}
      />
    </section>
  )
}
