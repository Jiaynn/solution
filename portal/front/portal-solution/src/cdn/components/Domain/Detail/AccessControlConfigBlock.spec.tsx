/**
 * @desc cases for access control config component
 * @email yaojingtian@qiniu.com
 */

import React from "react";

import { RendererUtils as Renderer } from "cdn/test";

import mockDomainDetail from "cdn/test/domain-detail-mock";

import AccessControlConfigBlock from "./AccessControlConfigBlock";

const domain = mockDomainDetail();

const mockHandle = jest.fn();

it("renders correctly", () => {
  const renderer = new Renderer();
  const tree = renderer
    .createWithAct(
      <AccessControlConfigBlock
        domain={domain}
        loading={false}
        handleConfigureReferer={mockHandle}
        handleConfigureTimeReferer={mockHandle}
        handleConfigureBsAuth={mockHandle}
        handleConfigureIpACL={mockHandle}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with loading: true", () => {
  const renderer = new Renderer();
  const tree = renderer
    .createWithAct(
      <AccessControlConfigBlock
        domain={domain}
        loading
        handleConfigureReferer={mockHandle}
        handleConfigureTimeReferer={mockHandle}
        handleConfigureBsAuth={mockHandle}
        handleConfigureIpACL={mockHandle}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
