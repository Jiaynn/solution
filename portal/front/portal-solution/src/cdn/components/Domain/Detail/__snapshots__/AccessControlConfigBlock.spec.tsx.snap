// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<section
  className="content-block"
>
  <h4
    className="content-block-title"
  >
    访问控制
  </h4>
  <div
    className="ant-table-wrapper content-config-table"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-table ant-table-default ant-table-scroll-position-left"
          style={Object {}}
        >
          <div
            className="ant-table-content"
          >
            <div
              className="ant-table-body"
              onScroll={[Function]}
              onWheel={[Function]}
              style={Object {}}
            >
              <table
                className=""
                style={Object {}}
              >
                <colgroup>
                  <col
                    style={
                      Object {
                        "minWidth": "120px",
                        "width": "120px",
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": "240px",
                        "width": "240px",
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": "180px",
                        "width": "180px",
                      }
                    }
                  />
                </colgroup>
                <thead
                  className="ant-table-thead"
                >
                  <tr
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            配置项
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            描述
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            当前配置
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            操作
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="Referer 防盗链"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      Referer 防盗链
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      配置 Request Header 中 referer 黑白名单，从而限制访问来源。
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="text-warning"
                      >
                        未开启
                      </span>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <button
                        className="ant-btn"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          修改配置
                        </span>
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="时间戳防盗链"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      时间戳防盗链
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <p>
                        设置密钥，配合签名过期时间来控制资源内容的访问时限。
                        <a
                          href="https://support.qiniu.com/hc/kb/article/195128/"
                          rel="noopener"
                          target="_blank"
                        >
                          帮助文档
                        </a>
                      </p>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="text-warning"
                      >
                        未开启
                      </span>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <button
                        className="ant-btn"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          修改配置
                        </span>
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="回源鉴权"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      回源鉴权
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <p>
                        配置回源验证策略，对 CDN 接收到的访问请求进行鉴权，适用于对防盗链有很高实时性要求的场景。
                        <a
                          href="https://developer.qiniu.com/fusion/manual/3930/back-to-the-source-authentication"
                          rel="noopener"
                          target="_blank"
                        >
                          帮助文档
                        </a>
                      </p>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="text-warning"
                      >
                        未开启
                      </span>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <button
                        className="ant-btn"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          修改配置
                        </span>
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="IP 黑白名单"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      IP 黑白名单
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <p>
                        允许或者禁止某些 IP 或 IP 段的访问，帮助您解决恶意 IP 盗刷、攻击等问题。
                      </p>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="text-warning"
                      >
                        未开启
                      </span>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <button
                        className="ant-btn"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          修改配置
                        </span>
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;

exports[`renders correctly with loading: true 1`] = `
<section
  className="content-block"
>
  <h4
    className="content-block-title"
  >
    访问控制
  </h4>
  <div
    className="ant-table-wrapper content-config-table"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div>
        <div
          className="ant-spin ant-spin-spinning ant-table-without-pagination ant-table-spin-holder"
        >
          <span
            className="ant-spin-dot ant-spin-dot-spin"
          >
            <i
              className="ant-spin-dot-item"
            />
            <i
              className="ant-spin-dot-item"
            />
            <i
              className="ant-spin-dot-item"
            />
            <i
              className="ant-spin-dot-item"
            />
          </span>
        </div>
      </div>
      <div
        className="ant-spin-container ant-spin-blur"
      >
        <div
          className="ant-table ant-table-default ant-table-scroll-position-left"
          style={Object {}}
        >
          <div
            className="ant-table-content"
          >
            <div
              className="ant-table-body"
              onScroll={[Function]}
              onWheel={[Function]}
              style={Object {}}
            >
              <table
                className=""
                style={Object {}}
              >
                <colgroup>
                  <col
                    style={
                      Object {
                        "minWidth": "120px",
                        "width": "120px",
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": undefined,
                        "width": undefined,
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": "240px",
                        "width": "240px",
                      }
                    }
                  />
                  <col
                    style={
                      Object {
                        "minWidth": "180px",
                        "width": "180px",
                      }
                    }
                  />
                </colgroup>
                <thead
                  className="ant-table-thead"
                >
                  <tr
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            配置项
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            描述
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            当前配置
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                    <th
                      className=""
                    >
                      <span
                        className="ant-table-header-column"
                      >
                        <div>
                          <span
                            className="ant-table-column-title"
                          >
                            操作
                          </span>
                          <span
                            className="ant-table-column-sorter"
                          />
                        </div>
                      </span>
                    </th>
                  </tr>
                </thead>
                <tbody
                  className="ant-table-tbody"
                >
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="Referer 防盗链"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      Referer 防盗链
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      配置 Request Header 中 referer 黑白名单，从而限制访问来源。
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="text-warning"
                      >
                        未开启
                      </span>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <button
                        className="ant-btn"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          修改配置
                        </span>
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="时间戳防盗链"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      时间戳防盗链
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <p>
                        设置密钥，配合签名过期时间来控制资源内容的访问时限。
                        <a
                          href="https://support.qiniu.com/hc/kb/article/195128/"
                          rel="noopener"
                          target="_blank"
                        >
                          帮助文档
                        </a>
                      </p>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="text-warning"
                      >
                        未开启
                      </span>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <button
                        className="ant-btn"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          修改配置
                        </span>
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="回源鉴权"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      回源鉴权
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <p>
                        配置回源验证策略，对 CDN 接收到的访问请求进行鉴权，适用于对防盗链有很高实时性要求的场景。
                        <a
                          href="https://developer.qiniu.com/fusion/manual/3930/back-to-the-source-authentication"
                          rel="noopener"
                          target="_blank"
                        >
                          帮助文档
                        </a>
                      </p>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="text-warning"
                      >
                        未开启
                      </span>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <button
                        className="ant-btn"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          修改配置
                        </span>
                      </button>
                    </td>
                  </tr>
                  <tr
                    className="ant-table-row ant-table-row-level-0"
                    data-row-key="IP 黑白名单"
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onDoubleClick={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    style={
                      Object {
                        "height": null,
                      }
                    }
                  >
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="ant-table-row-indent indent-level-0"
                        style={
                          Object {
                            "paddingLeft": "0px",
                          }
                        }
                      />
                      IP 黑白名单
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <p>
                        允许或者禁止某些 IP 或 IP 段的访问，帮助您解决恶意 IP 盗刷、攻击等问题。
                      </p>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <span
                        className="text-warning"
                      >
                        未开启
                      </span>
                    </td>
                    <td
                      className=""
                      onClick={[Function]}
                    >
                      <button
                        className="ant-btn"
                        onClick={[Function]}
                        type="button"
                      >
                        <span>
                          修改配置
                        </span>
                      </button>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>
`;
